sdfc-opportunity-to-sap-sales-order/main.bal

```diff
@@ -0,0 +1,113 @@
+const int HTTP_STATUS_OK = 200;
+const int HTTP_STATUS_CREATED = 201;
+const string HEADER_KEY_CSRF_TOKEN = "x-csrf-token";
+const string HEADER_FETCH_VALUE = "fetch";
+
+final http:Client sapHttpClient = check getSAPHttpClient();
+
+configurable SalesforceListenerConfig salesforceListenerConfig = ?;
+
+configurable SAPAuthConfig sapAuthConfig = ?;
+
+listener salesforce:Listener sfdcEventListener = new ({
+    username: salesforceListenerConfig.username,
+    password: salesforceListenerConfig.password,
+    channelName: CHANNEL_NAME,
+    environment: ENVIRONMENT
+});
+
+service salesforce:RecordService on sfdcEventListener {
+    isolated remote function onCreate(salesforce:EventData payload) returns error? {
+        log:printInfo("New opportunity is created!");
+        check createSAPSalesOrder();
+    }
+
+    isolated remote function onUpdate(salesforce:EventData payload) returns error? {
+        return;
+    }
+
+    isolated remote function onDelete(salesforce:EventData payload) returns error? {
+        return;
+    }
+
+    isolated remote function onRestore(salesforce:EventData payload) returns error? {
+        return;
+    }
+}
+
+isolated function getSAPHttpClient() returns http:Client|error {
+    http:CredentialsConfig basicAuthHandler = {username: sapAuthConfig.username, password: sapAuthConfig.password};
+    return new (url = SAP_URL, auth = basicAuthHandler, cookieConfig = {enabled: true});
+}
+
+isolated function createSAPSalesOrder() returns error? {
+    json request = {
+        "PurchaseOrderType": "NB",
+        "Supplier": "17300096",
+        "PurchasingOrganization": "1710",
+        "PurchasingGroup": "001",
+        "CompanyCode": "1710",
+        "_PurchaseOrderItem": [
+            {
+                "Plant": "1710",
+                "OrderQuantity": 10,
+                "PurchaseOrderQuantityUnit": "kg",
+                "AccountAssignmentCategory": "U",
+                "Material": "E001",
+                "NetPriceAmount": 100,
+                "DocumentCurrency": "USD",
+                "TaxJurisdiction": "KY00000000"
+            }
+        ]
+    };
+    map<string|string[]> headerParams = check getHeaderParams();
+    http:Response|http:ClientError response = sapHttpClient->post(
+        path = SAP_REQUEST_PATH,
+        message = request,
+        headers = headerParams
+        );
+    if response is http:Response {
+        if response.statusCode == HTTP_STATUS_CREATED {
+            json responseBody = check response.getJsonPayload();
+            string purchaseOrderId = check responseBody.PurchaseOrder;
+            log:printInfo("Successfully created an SAP purchase order with id: " + purchaseOrderId);
+            return;
+        }
+        log:printError("Error: " + check response.getTextPayload());
+        return;
+    }
+    log:printError("Error: " + response.message());
+}
+
+isolated function getHeaderParams() returns map<string|string[]>|error {
```

Comment:
Why are we returning the http response in this `getHeaderParams` method?

Reply:
I think what you need is to extract the CSRF token and send it with the sales order request.